sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -171,8 +171,26 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(substitutedProjection, child)
 
+    case gen@Generate(_: Explode, _, _, _, c) =>
```

Comment:
Oh, I see.  Sorry I did not look closely enough.  In general we do this by matching on a project that is above the node in question.  This works because projects will get pushed down or created in case where unneeded attributes exist.  There are other examples in this file where we do this.